src/extensions/AssetMigration.sol

```diff
@@ -0,0 +1,250 @@
+   */
+  function _getWrappedNativeToken() internal view returns (IWETH) {
+    return _getAssetMigration()._wnt;
+  }
+
+  /**
+   * @dev Migrates the given tokens to the specified addresses.
+   *
+   * When the token is the native (i.e RON or ETH), it will be wrapped (i.e WRON, WETH).
+   *
+   * Requirements:
+   * - The caller must be the migrator.
+   * - The length of the arrays must be the same.
+   * - The length of the arrays must not be zero.
+   */
+  function migrateERC20(address[] calldata tokens, uint256[] calldata amounts) external onlyMigrator validInput(_toUint256s(tokens), amounts) {
+    uint256 length = amounts.length;
+    IERC20 token;
+
+    for (uint256 i; i < length; ++i) {
+      token = IERC20(tokens[i]);
+
+      if (address(token) == address(_NATIVE_TOKEN_INDICATOR)) {
+        token = _wrap(amounts[i]);
+      }
+
+      address recipient = _requireWhitelisted(address(token));
+      token.safeTransfer(recipient, amounts[i]);
+    }
+  }
+
+  /**
+   * @dev Migrates the given ERC721 tokens to the specified addresses.
+   *
+   * Requirements:
+   * - The caller must be the migrator.
+   * - The length of the arrays must be the same.
+   * - The length of the arrays must not be zero.
+   * - If `recipient` is not whitelisted and not inherit from the `ERC721Receiver` interface, it will revert.
+   */
+  function migrateERC721(address[] calldata tokens, uint256[] calldata ids) external onlyMigrator validInput(_toUint256s(tokens), ids) {
+    uint256 length = tokens.length;
+
+    for (uint256 i; i < length; ++i) {
+      address recipient = _requireWhitelisted(tokens[i]);
+      IERC721(tokens[i]).safeTransferFrom(address(this), recipient, ids[i]);
+    }
+  }
+
+  /**
+   * @dev Whitelists the recipients for the given tokens.
+   *
+   * Requirements:
+   * - Must go through proposal via `BridgeManager`.
+   */
+  function whitelist(address[] calldata tokens, address[] calldata recipients) external onlyProxyAdmin validInput(_toUint256s(recipients), _toUint256s(tokens)) {
+    _whitelist(recipients, tokens);
+  }
+
+  /**
+   * @dev Renounces the migrator role.
+   */
+  function renounceMigrator() external onlyMigrator {
+    _setMigrator(address(0));
+  }
+
+  /**
+   * @dev Get all whitelisted addresses for the given tokens.
+   */
+  function getWhitelistedAddresses(
+    address[] calldata tokens
+  ) external view returns (address[] memory whitelisteds) {
+    AssetMigrationStorage storage $ = _getAssetMigration();
+
+    uint256 length = tokens.length;
+    whitelisteds = new address[](length);
+
+    for (uint256 i; i < length; ++i) {
+      whitelisteds[i] = $._whitelist[tokens[i]];
+    }
+  }
+
+  /**
+   * @dev Returns the pointer of the AssetMigrationStorage struct.
+   */
+  function _getAssetMigration() private pure returns (AssetMigrationStorage storage $) {
+    bytes32 loc = _$$AssetMigrationLocation();
+    require(loc != 0, ErrInvalidStorageLocation());
```

Comment:
`_$$AssetMigrationLocation()` not needed to be virtual (or exist at all) and so this validation. Can we put the slot value here?

Reply:
I think exposing into internal func may support us for writing unit test

Reply:
removed.